### ISSUE TITLE
Cannot create new TCP server after network disconnect

### ISSUE THREAD

**Issue by kaincenteno** [diasurgical/devilutionX#132]
# Issue

After a game is disconnected the game won't allow you to create a new one because it doesn't release the port. A work around is to simply restart the game.

=========================================

# Original description:

Trying to connect multiplayer TCP, ports are forwarded properly. Error message on client side says its not running version 0.2.0 (both client and server are running 0.4.0 windows version)

After error message is shown on the client's end, the server becomes unresponsive and cannot ALT+F4 or ALT+TAB, server computer has to wait 5 minutes for devilutionx to close automatically.

If there's any info more needed please let me know and I can provide it 😸

Below debug:

```DUMMY: SetUnhandledExceptionFilter @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:507
DUMMY: GetSystemInfo @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:573
DUMMY: CreateFileMappingA @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:665
DUMMY: FindWindowA : class: DIABLO window: (null)
DUMMY: LoadIconA @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:431
DUMMY: LoadCursorA @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:446
DUMMY: LoadImageA @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:440
DUMMY: RegisterClassExA @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:399
DUMMY: UpdateWindow @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:382
DUMMY: UiAppActivate @ /home/ajenbo/code/diablo/devilutionX/SourceX/DiabloUI/diabloui.cpp:422
DUMMY: GetWindowLongA @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:778
DUMMY: SetWindowLongA @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:784
DUMMY: SDrawRealizePalette @ /home/ajenbo/code/diablo/devilutionX/SourceX/storm/storm.cpp:705
DUMMY: DefWindowProcA @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:769
DUMMY: SetCooperativeLevel @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/ddraw.cpp:100
DUMMY: SetDisplayMode @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/ddraw.cpp:106
DUMMY: CreateSurface @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/ddraw.cpp:57
DUMMY: CreatePalette @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/ddraw.cpp:42
DUMMY: SetPalette @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/ddraw.cpp:203
DUMMY: GdiSetBatchLimit @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:270
DUMMY: GetCaps @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/ddraw.cpp:159
DUMMY: CreateSurface @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/ddraw.cpp:57
DUMMY: SDrawManualInitialize @ /home/ajenbo/code/diablo/devilutionX/SourceX/storm/storm_dx.cpp:19
DUMMY: WaitForVerticalBlank @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/ddraw.cpp:116
DUMMY: BltFast @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/ddraw.cpp:136
DUMMY: Unlock @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/ddraw.cpp:209
DUMMY: sound_DirectSoundCreate @ /home/ajenbo/code/diablo/devilutionX/SourceX/sound.cpp:409
DUMMY: SVidInitialize @ /home/ajenbo/code/diablo/devilutionX/SourceX/storm/storm.cpp:424
DUMMY: GetFileVersionInfoSizeA @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:167
DUMMY: GetFileVersionInfoA @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:175
DUMMY: VerQueryValueA @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:183
BasePath: D:\devilutionx-windows-i386\
DUMMY: SFileSetBasePath @ /home/ajenbo/code/diablo/devilutionX/SourceX/storm/storm.cpp:699
BasePath: D:\devilutionx-windows-i386\
DUMMY: SFileSetBasePath @ /home/ajenbo/code/diablo/devilutionX/SourceX/storm/storm.cpp:699
DUMMY: sound_CreateSoundBuffer @ /home/ajenbo/code/diablo/devilutionX/SourceX/sound.cpp:274
DUMMY: InvalidateRect @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:373
DUMMY: UpdateWindow @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:382
DUMMY: PeekMessageA : unknown SDL message 0x1100
DUMMY: false_avail : return false although event avaliable
DUMMY: PeekMessageA : unknown SDL message 0x1100
DUMMY: false_avail : return false although event avaliable
DUMMY: PeekMessageA : unknown SDL message 0x1100
DUMMY: false_avail : return false although event avaliable
DUMMY: PeekMessageA : unknown SDL message 0x1100
DUMMY: false_avail : return false although event avaliable
DUMMY: PeekMessageA : unknown SDL message 0x1100
DUMMY: false_avail : return false although event avaliable
DUMMY: PeekMessageA : unknown SDL message 0x1100
DUMMY: false_avail : return false although event avaliable
DUMMY: PeekMessageA : unknown SDL message 0x1100
DUMMY: false_avail : return false although event avaliable
DUMMY: false_avail : return false although event avaliable
DUMMY: false_avail : return false although event avaliable
DUMMY: false_avail : return false although event avaliable
DUMMY: false_avail : return false although event avaliable
DUMMY: false_avail : return false although event avaliable
DUMMY: PeekMessageA : unknown SDL message 0x302
DUMMY: false_avail : return false although event avaliable
DUMMY: false_avail : return false although event avaliable
DUMMY: DispatchMessageA @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc_msg.cpp:284
DUMMY: false_avail : return false although event avaliable
DUMMY: false_avail : return false although event avaliable
DUMMY: false_avail : return false although event avaliable
DUMMY: PostMessageA @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc_msg.cpp:294
DUMMY: SDrawGetFrameWindow @ /home/ajenbo/code/diablo/devilutionX/SourceX/storm/storm.cpp:307
DUMMY: UiProfileGetString @ /home/ajenbo/code/diablo/devilutionX/SourceX/DiabloUI/diabloui.cpp:386
DUMMY: OpenFile @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:240
DUMMY: OpenFile @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:240
DUMMY: OpenFile @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:240
DUMMY: OpenFile @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:240
DUMMY: OpenFile @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:240
DUMMY: OpenFile @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:240
DUMMY: OpenFile @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:240
DUMMY: OpenFile @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:240
DUMMY: OpenFile @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:240
DUMMY: OpenFile @ /home/ajenbo/code/diablo/devilutionX/SourceX/miniwin/misc.cpp:240
DUMMY: UiSetupPlayerInfo @ /home/ajenbo/code/diablo/devilutionX/SourceX/DiabloUI/diabloui.cpp:394
DUMMY: SNetDestroy @ /home/ajenbo/code/diablo/devilutionX/SourceX/storm/storm_net.cpp:61
```
<img width="480" alt="error" src="https://user-images.githubusercontent.com/26943220/59824151-235d6480-92e5-11e9-814e-59a96903a21c.PNG">

**Comment by AJenbo** [comment]
Thanks for reporting, I'm assuming both are running the official 32bit release and not a self-build version?

The message should probably be made more clear, it requires 0.2.0 or higher.

Could you test with both the server and client running on the same machine (disable fullscreen and input grab in diablo.ini).
If you could test with 0.3.1 and 0.2.0 that would also be helpful.

**Comment by kaincenteno** [comment]
Hi @AJenbo 

Yes we are running the official 32bit version download from the readme.md 😸

Sure we'll test on that same machine, let me explain a bit more about the issue (as it was experienced on 2 instances)

**The first Instance (I was the TCP server)**

Me, my fiancee, and a friend were playing together. (All running the official 32bit windows 0.4.0 version connected through UDP)

Everything was fine until my friends headphones disconnected, which dropped him back to the windows Desktop on a weird resolution and unable to do anything so he ALT+F4 the game.

Once he was trying to connect again to the Server, he got that pop up message complaining about version 0.2.0 not detected. So he went again and redownloaded it and afterwards was able to connect (even tho the version he downloaded was the same version he was using before, not sure if there's a configuration file that gets created on that folder outside $HOME).

**The Second Instance (My Fiancee was the Server)**

He started the server, made sure his router + firewall were routing properly to his computer beforehand. I was able to access that machine with pings and all that (we are playing over the internet not LAN).

He was fine on his server until the moment i was trying to make a connection with him, the moment I did his game frozed and he was unable to ALT + F4 or do anything, he had to wait 5 minutes for the application to close/window crash. On my end I received that message that my friend was getting before on the first instance, i did not try doing the workaround he did but will definitely will try tonight. 

Sorry for the wall of text, just wanted to give you more info, I'll get the info you requested when back home😸

**Comment by AJenbo** [comment]
Thanks for the extra info, it's possible that the game isn't releasing the port when it exits unexpected, a restart should fix that, but hopefully we can also improve it.

**Comment by kaincenteno** [comment]
Happy monday @AJenbo 

Sorry for the late response. It ended up being User Error, I know hella embarrasing, had to troubleshoot using wireshark, his router had this weird configuration that the firewall was assigned using hostname instead of IP address, I like to do it oldschool so asked him to remove the hostname and add the rule using good old IP address. After that was done everything worked fine.

But one thing we noticed is that UDP doesn't work for any of us, even when trying to connecting to 127.0.0.1 locally on each machine. If we try to do that, the client will kick the server out of the game. (not sure if UDP is already implemented on devilutionX)

About the issue with my friend not being able to connect back to the server, that we haven't tried as I would have to force shutdown my machine, but will try to reproduce it tonight :)

**Comment by AJenbo** [comment]
Good to hear that you were able to solve the issue. UDP was the first to be implemented as it resembles the original IPX more, but maybe an issue has crept into it since we mainly test the TCP version nowadays.

**Comment by kaincenteno** [comment]
@AJenbo I was able to reproduce the issue of the pop up asking for version 0.2.0 so it was like you mentioned of the game not releasing the port.

I was playing, the fiancee was the server and I was the client.

My roommate decided to powercycle the router so I lost connection. The game didn't close but it kicked me out back to the menu screen. Once back online I tried connecting to the fiance's game and I was unable to and received the error of version 0.2.0

I closed the game, and logged in again without issues.

Let me know if there's anything needed, i'd like to help as much as I can 😸
